src/warpychain/initial_syncer.go

```diff
@@ -0,0 +1,258 @@
+
+	"github.com/warp-contracts/syncer/src/utils/config"
+	"github.com/warp-contracts/syncer/src/utils/monitoring"
+	"github.com/warp-contracts/syncer/src/utils/task"
+	"gorm.io/gorm"
+)
+
+type InitialSyncer struct {
+	db *gorm.DB
+	*task.Task
+	warpyChainClient *WarpyChainClient
+
+	maxElapsedTime time.Duration
+	maxInterval    time.Duration
+	monitor        monitoring.Monitor
+}
+
+func NewInitialSyncer(config *config.Config) (self *InitialSyncer) {
+	self = new(InitialSyncer)
+
+	self.Task = task.NewTask(config, "initial_syncer").
+		WithWorkerPool(config.WarpyChainer.InitialSyncerNumWorkers, config.WarpyChainer.InitialSyncerWorkerQueueSize)
+
+	return
+}
+
+func (self *InitialSyncer) WithDB(db *gorm.DB) *InitialSyncer {
+	self.db = db
+	return self
+}
+
+func (self *InitialSyncer) WithMonitor(monitor monitoring.Monitor) *InitialSyncer {
+	self.monitor = monitor
+	return self
+}
+
+func (self *InitialSyncer) WithBackoff(maxElapsedTime, maxInterval time.Duration) *InitialSyncer {
+	self.maxElapsedTime = maxElapsedTime
+	self.maxInterval = maxInterval
+	return self
+}
+
+func (self *InitialSyncer) WithWarpyChainClient(client *WarpyChainClient) *InitialSyncer {
+	self.warpyChainClient = client
+	return self
+}
+
+func (self *InitialSyncer) Sync() (err error) {
+	ctx, cancel := context.WithTimeout(self.Ctx, self.Config.WarpyChainer.PollerTimeout)
+	defer cancel()
+
+	var lastSyncedHeight int64
+	var currentBlockHeight int64
+	var currentBlockHash string
+
+	err = task.NewRetry().
+		WithContext(self.Ctx).
+		WithMaxElapsedTime(self.maxElapsedTime).
+		WithMaxInterval(self.maxInterval).
+		WithAcceptableDuration(self.maxInterval * 2).
+		WithOnError(func(err error, isDurationAcceptable bool) error {
+			if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+				return backoff.Permanent(err)
+			}
+
+			self.Log.WithError(err).Warn("Could not get last synced block info, retrying...")
+			return err
+		}).
+		Run(func() error {
+			lastSyncedHeight, err = self.warpyChainClient.getLastSyncedBlockHeight(ctx)
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get last synced height")
+				return err
+			}
+
+			currentBlockHeight, currentBlockHash, err = self.warpyChainClient.getCurrentBlockHeight()
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get current block height")
+				return err
+			}
+
+			return err
+		})
+
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get last synced block info")
+		// it requires an alert
+		self.monitor.GetReport().WarpyChainer.Errors.InitialSyncerLastSyncedError.Inc()
+		return
+	}
+
+	self.Log.WithField("lastSyncedHeight", lastSyncedHeight).WithField("currentBlockHeight", currentBlockHeight).
+		WithField("blocksToSync", currentBlockHeight-lastSyncedHeight).Info("Number of blocks to sync")
+
+	for ok := true; ok; ok = lastSyncedHeight < currentBlockHeight-1 {
+		blocksBatchSize := 20
+		if currentBlockHeight-lastSyncedHeight < 20 {
```

Comment:
I believe, with the help of recently added https://medium.com/gopher-time/go-1-21-new-built-ins-min-max-clear-explored-cda29ee56f22 , we could write here sth like
`blocksBatchSize := min(20, currentBlockHeight - lastSyncedHeight)`